### DISCUSSION TITLE
404 when running on apache

### DISCUSSION THREAD

**Discussion by juanbautista0** [clue/framework-x#164] [Q&A] [answered]
I am testing apache2 ubuntu and I get error 404. I have configured the htaccess as recommended but still the error message persists.

The same thing happened to me testing the Slim frame and with a method where the base path was defined, it was solved.

```
$app->setBasePath('/test');
```

What do you do here ?

The configuration is as follows:

```
acme/
├── public/
│   ├── .htaccess
│   └── index.php
├── vendor/
├── composer.json
└── composer.lock
```
path: /.htaccess

``` title="public/.htaccess"
RewriteEngine on
RewriteRule ^$ public/ [L]
RewriteRule (.*) public/$1 [L]

```
path: public/.htaccess

``` title="public/.htaccess"
RewriteEngine On

RewriteCond %{REQUEST_FILENAME} !-d
RewriteCond %{REQUEST_FILENAME} !-f
RewriteRule .* index.php

# This adds support for authorization header
SetEnvIf Authorization .+ HTTP_AUTHORIZATION=$0
```

**Comment by juanbautista0** [comment, marked as answer]
I want to share the solution I found for my problem. If you want to deploy in apache it is recommended to use a virtual host and in production environment a subdomain, so it will work with the base path.

Thanks

**Reply by clue** [reply]
@juanbautista0 Happy to hear you've got this sorted out already! I'm curious, may I ask you to share what was wrong about your configuration? Perhaps we can get some additional insights and improve our documentation for future users 👍

Apache is known to work just fine as documented and we're not currently aware of any limitations whatsoever: https://framework-x.org/docs/best-practices/deployment/#apache (see also discussion #111)

**Reply by RebafC** [reply]
I had the same issue. I am on Win 10 using a Neard WAMP stack. The example page from the quickstart would load using php built in server but failed using apache. But once I defined a virtual host the page loaded without error.

**Reply by clue** [reply]
@RebafC May I ask you to share your virtual host configuration? Perhaps we can take a look at see if we can improve our documentation. PRs also very much welcome!

**Reply by RebafC** [reply]
[framework-x.local.txt](https://github.com/clue/framework-x/files/9081433/framework-x.local.txt)

**Reply by juanbautista0** [reply]
My file is:

```
<VirtualHost *:80>
    ServerAdmin admin@myapp.com
    ServerName myapp.dev.com
    ServerAlias www.myapp.dev.com
    DocumentRoot /var/www/myapp.dev.com/cloud
    ErrorLog ${APACHE_LOG_DIR}/error.log
    CustomLog ${APACHE_LOG_DIR}/access.log combined
</VirtualHost>
```

Summary

I think the problem is that it cannot be placed in a subdirectory that is not root, so the htaccess file is directing the traffic to the public directory will recognize as base path the full path.

❌: mydomain.dev/app-framework-x
👌: mydomain.dev

In the case of Slim framework it does a url mapping and determines the base path, you can also set it manually. 

```
$app->setBasePath('/myapp');
```
It would be very positive if this could be dynamic or user definable. 

Thank you very much for your attention and congratulations for this great tool that is Framework-x.

**Reply by SimonFrings** [reply]
@juanbautista0 Really glad you like the project 👍 

There is also the possibility to use a middleware for this. We opened up a new discussion to add [PSR-15 compatibility](https://github.com/clue/framework-x/discussions/185), which gives us a whole new world of opportunities.

Either way, it sounds like a useful addition to the project, we also have a discussion about [route groups](https://github.com/clue/framework-x/discussions/134) which could also be a solution for this.
